google/cloud/bigtable/batcher.py

```diff
@@ -13,17 +13,117 @@
 # limitations under the License.
 
 """User friendly container for Google Cloud Bigtable MutationBatcher."""
+import threading
+import queue
+import concurrent.futures
+import atexit
 
+from google.api_core.exceptions import from_grpc_status
 
-FLUSH_COUNT = 1000
+
+# Max number of items in the queue. Queue will be flushed if this number is reached
+FLUSH_COUNT = 100
+
+# Max number of mutations for a single row at one time
 MAX_MUTATIONS = 100000
-MAX_ROW_BYTES = 5242880  # 5MB
+
+# Max size (in bytes) for a single mutation
+MAX_ROW_BYTES = 20 * 1024 * 1024  # 20MB
+
+# Max size (in bytes) for a single request
+MAX_MUTATIONS_SIZE = 100 * 1024 * 1024  # 100MB
+
+
+class MutationsBatchError(Exception):
+    """Error in the batch request"""
+
+    def __init__(self, message, exc):
+        self.exc = exc
+        self.message = message
+        super().__init__(self.message)
 
 
 class MaxMutationsError(ValueError):
     """The number of mutations for bulk request is too big."""
 
 
+class BatcherIsClosedError(ValueError):
+    """Batcher is already closed and not accepting any mutation."""
+
+
+class _MutationsBatchQueue(object):
+    """Private Threadsafe Queue to hold rows for batching."""
+
+    def __init__(self, max_row_bytes=MAX_ROW_BYTES, flush_count=FLUSH_COUNT):
+        """Specify the queue constraints"""
+        self._queue = queue.Queue(maxsize=flush_count)
+        self.total_mutation_count = 0
+        self.total_size = 0
+        self.max_row_bytes = max_row_bytes
+
+    def get(self, block=True, timeout=None):
+        """Retrieve an item from the queue. Recalculate queue size."""
+        row = self._queue.get(block=block, timeout=timeout)
+        mutation_size = row.get_mutations_size()
+        self.total_mutation_count -= len(row._get_mutations())
+        self.total_size -= mutation_size
+        return row
+
+    def put(self, item, block=True, timeout=None):
+        """Insert an item to the queue. Recalculate queue size."""
+
+        mutation_count = len(item._get_mutations())
+        mutation_size = item.get_mutations_size()
+
+        if mutation_count > MAX_MUTATIONS:
+            raise MaxMutationsError(
+                "The row key {} exceeds the number of mutations {}.".format(
+                    item.row_key, mutation_count
+                )
+            )
+
+        if mutation_size > MAX_MUTATIONS_SIZE:
+            raise MaxMutationsError(
+                "The row key {} exceeds the size of mutations {}.".format(
+                    item.row_key, mutation_size
+                )
+            )
+
+        self._queue.put(item, block=block, timeout=timeout)
+
+        self.total_size += item.get_mutations_size()
+        self.total_mutation_count += mutation_count
+
+    def full(self):
+        """Check if the queue is full."""
+        if (
+            self.total_mutation_count >= MAX_MUTATIONS
+            or self.total_size >= self.max_row_bytes
+            or self._queue.full()
```

Comment:
Why do we need to check if the queue is full? In the batcher code, we flush when the Batcher is full:
```
if self._rows.full():
    self.flush_async()
```
I'm not sure if the behavior is still correct? 

I think full() should only check for 2 things: 
- Number of elements reached the batch element threshold
- Number of bytes reached the batch bytes threshold

The queue is only be used to block user from adding more elements when it's full. We don't trigger another flush when the queue is full.